### PR TITLE
Verify the state root exists before creating a State

### LINKED ISSUE #1212: Trie root not found!
reported by @jacogr 

```
~ parity --testnet --rpc
2016-06-02 18:13:21 Using a conversion rate of Ξ1 = US$13.9 (17129155000 wei/gas)
2016-06-02 18:13:22 Starting Parity/v1.2.0-unstable-35753f2-20160602/x86_64-macos/rustc1.9.0
2016-06-02 18:13:22 Configured for Morden using "Ethash" engine
2016-06-02 18:13:22 Public node URL: enode://1f64f15db9f842087c938c6b1bcc2719595e1bd7fa9fc1dae38a89f31fce14cc67e50aa6e3939106ac987174100edab6fc50900716b970782c56aea4e2012c5a@0.0.0.0:30303
TrieDB::new(3f56…626f): Trie root not found!
thread 'hyper-server' panicked at 'Trie root not found!', util/src/trie/triedb.rs:64
note: Run with `RUST_BACKTRACE=1` for a backtrace.
2016-06-02 18:13:30 Finishing work, please wait...
thread '<main>' panicked at 'called `Result::unwrap()` on an `Err` value: Closed', ../src/libcore/result.rs:746
thread '<main>' panicked at 'called `Result::unwrap()` on an `Err` value: Any', ../src/libcore/result.rs:746
stack backtrace:
   1:        0x101670358 - std::sys::backtrace::tracing::imp::write::h4c73fcd3363076f5
   2:        0x101673975 - std::panicking::default_hook::_$u7b$$u7b$closure$u7d$$u7d$::h0422dbb3077e6747
   3:        0x1016735ae - std::panicking::default_hook::haac48fa641db8fa2
   4:        0x10165dbd6 - std::sys_common::unwind::begin_unwind_inner::h39d40f52add53ef7
   5:        0x10165ea5e - std::sys_common::unwind::begin_unwind_fmt::h64c0ff793199cc1b
   6:        0x10166f627 - rust_begin_unwind
   7:        0x10169a920 - core::panicking::panic_fmt::h73bf9d7e8e891a73
   8:        0x10110a2b0 - core::result::unwrap_failed::h0bdd56634916bdf6
   9:        0x10110a112 - _<server..Listening as std..ops..Drop>::drop::hf7eefae0ab712359
  10:        0x10110a508 - hyper::server::Listening::close::h08ed0aabe12182f0
  11:        0x1010f1f3e - _<Server as std..ops..Drop>::drop::h612aedc21573f07e
  12:        0x100e1a1a7 - parity::execute_client::h93c52c469cd553fa
  13:        0x100e07eb4 - parity::execute::h4228f2b3aef67c46
  14:        0x100e063df - parity::main::hc0b8b9a4c77213f8
  15:        0x1016731a2 - std::sys_common::unwind::try::try_fn::h09ba69fd13531e58
  16:        0x10166f5bb - __rust_try
  17:        0x101672fe4 - std::rt::lang_start::h5b0863080165c75e
thread panicked while panicking. aborting.
[1]    27726 illegal hardware instruction  parity --testnet --rpc
```

crashed by calling rpc method:

```
{"jsonrpc":"2.0","method":"eth_getBalance","params":["0x63cf90d3f0410092fc0fca41846f596223979195","0x65432"],"id":6}
```

### PR DESCRIPTION
**Pull request by rphmeier**
Should fix #1212

### CODE REVIEW

File: util/src/journaldb/traits.rs [hunk 1]
```diff
@@ -42,4 +42,7 @@ pub trait JournalDB : HashDB + Send + Sync {
 	fn state(&self, _id: &H256) -> Option<Bytes> {
 		None
 	}
+
+	/// Whether this database does pruning.
+	fn does_pruning(&self) -> bool { true }
```
Comment (NikVolf) [inline util/src/journaldb/traits.rs]: `is_pruning` or even `pruning` better?

Comment (gavofyork) [inline util/src/journaldb/traits.rs]: `is_pruned`

File: ethcore/src/client/client.rs [hunk 2]
```diff
@@ -371,9 +371,24 @@ impl<V> Client<V> where V: Verifier {
...
+
+		self.block_header(id).and_then(|header| {
 			let db = self.state_db.lock().unwrap().boxed_clone();
-			State::from_existing(db, HeaderView::new(&header).state_root(), self.engine.account_start_nonce())
+
+			let root = HeaderView::new(&header).state_root();
+			// check that the block is not too old -- blocks within `HISTORY` blocks of the best will
+			// always be available. If the block could be too old, check if its state root is valid.
+			if db.does_pruning()
+			&& self.chain.best_block_number() >= block_number + HISTORY
```
Comment (gavofyork) [inline ethcore/src/client/client.rs]: bad style - run-on lines should be indented. in this case, i would place `let is_old = self.chain.best_block_number() >= block_number + HISTORY;` on its own line above and keep the `if` condition on a single line.

File: ethcore/src/client/client.rs [hunk 3]
```diff
@@ -371,9 +371,24 @@ impl<V> Client<V> where V: Verifier {
...
+			&& !db.contains(&root) {
```
Comment (gavofyork) [inline ethcore/src/client/client.rs]: why not just `if !db.contains(&root)`?

Comment (rphmeier) [inline ethcore/src/client/client.rs]: my intuition here was that a db query is relatively expensive, so the other checks are in order to short-circuit the expression

**Pull request merged**